src/metabase/query_processor/preprocess.clj

```diff
@@ -182,19 +191,23 @@
        ([query middleware-fn]
         (try
           (assert (ifn? middleware-fn))
-          ;; make sure the middleware returns a valid query... this should be dev-facing only so no need to i18n
-          (u/prog1 (middleware-fn query)
-            (qp.debug/debug>
-              (when-not (= <> query)
-                (let [middleware-fn-name (middleware-fn-name middleware-fn)]
-                  (list middleware-fn-name '=> <>
-                        ^{:portal.viewer/default :portal.viewer/diff}
-                        [(or (-> <> meta :converted-form) query)
-                         <>]))))
+          (let [start-timer (u/start-timer)]
             ;; make sure the middleware returns a valid query... this should be dev-facing only so no need to i18n
-            (when-not (map? <>)
-              (throw (ex-info (format "Middleware did not return a valid query.")
-                              {:fn (middleware-fn-name middleware-fn), :query query, :result <>, :type qp.error-type/qp}))))
+            (u/prog1 (middleware-fn query)
+              (let [duration-ms (u/since-ms start-timer)]
+                (when (> duration-ms slow-middleware-warning-threshold-ms)
+                  (log/warnf "Slow middleware: %s took %s" (middleware-fn-name middleware-fn) (u/format-milliseconds duration-ms))))
+              (qp.debug/debug>
+                (when-not (= <> query)
+                  (let [middleware-fn-name (middleware-fn-name middleware-fn)]
```

Comment:
having this duration in the tap output was very helpful to quickly debug stats. think it should go there?

Reply:
one issue is that log messages are intermingled from threads. the taps created when binding `metabase.query-processor.debug/*debug*` to true are all from a single thread. but it's not necessary. just an idea. I added the timestamp to the tap handler and it was fine. so i guess that's my answer

Reply:
I think slow middleware warning messages being intermingled is fine, we just want to know if any middleware is being slow ever and it doesn't really matter which query triggered it